src/forms-pipe.js

```diff
@@ -121,6 +122,20 @@ export async function formsPipe(state, req) {
   const { path } = state.info;
   const resourcePath = `${path}.json`;
 
+  let body;
+  try {
+    body = await extractBodyData(req);
+  } catch (err) {
+    return error(log, err.message, 400, res);
+  }
+
+  // verify captcha if needed
+  try {
+    await validateCaptcha(state, body);
+  } catch (e) {
+    return error(log, e.message, e.code, res);
+  }
+
```

Comment:
@hannessolo can you move this after the check below? this avoids extra processing for invalid posts.

Reply:
Sure - done.